keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1329 @@
+- Watches Leader Lease, waiting to be elected leader by the Coordinated Election
+  Controller
+- When it becomes leader:
+  - Perform role of active component instance
+  - Renew leader lease periodically
+  - Stop renewing if lease is marked `coordination.k8s.io/end-of-term: true`
+- If leader lease expires:
+  - Shutdown (yielding leadership) and restart as a candidate component instance
+
+```mermaid
+flowchart TD
+    A[Started] -->|Create LeaseCandidate Lease| B
+    B[Candidate] --> |Elected| C[Leader]
+    C --> |Renew Leader Lease| C
+    C -->|End of Term / Leader Lease Expired| D[Shutdown]
+    D[Shutdown] -.-> |Restart| A
+```
+
+### Renewal Interval and Performance
+The leader lease will have similar renewal interval and duration (2s and 15s).
+
+For component leases, keeping a short renewal interval will add many unnecessary writes to the apiserver.
+The component leases renewal interval will default to 5 mins.
+
+When the leader lease is marked as end of term or available, the coordinated leader election controller will
+add an annotation to all component leases `coordination.k8s.io/pending-ack` and wait up to 5 seconds.
+During that time, components must update their component lease to remove the annotation.
+The leader election controller will then pick the leader based on its criteria from the set of component leases that have ack'd the request.
+
+### Strategy
+
+Coordinated Leader Election allows the coordinated leader election controller to
+determine the most eligible leader out of a set of LeaseCandidate objects. This
+is a feature that could be extended to other use cases if a specific selection
+strategy is required. By default, CLE will default to the lowest compatibility
+version to respect [Kubernetes component version skew
+policy](https://kubernetes.io/releases/version-skew-policy/).
+
+Strategy is an option specified by a `LeaseCandidate` object, in most cases the
+controller responsible renewing the `LeaseCandidate` lease. The value for the
+strategy between different `LeaseCandidate` objects leading the same `Lease`
+should be the same, but during mixed version states, there is a possibility that
+they may differ. We will use a consensus protocol that favors the algorithm with
+the highest priority. The priority is a fixed list that is predetermined. For
+now, this is `NoCoordination` > `MinimumCompatibilityVersion`. For example, if
+three `LeaseCandidate` objects exist and two objects select
+`MinimumCompatibilityVersion` while the third selects `NoCoordination`,
+`NoCoordination` will take precedent and the coordinated leader election
+controller will use `NoCoordination` as the election strategy. The final
+strategy used will be written to the `Lease` object when the CLE controller
+creates the `Lease` for a suitable leader. This has the benefit of providing
+better debugging information and allows short circuiting of an election if the
+set of candidates and selected strategy is the same as before.
```

Comment:
Could we provide more detail on the alternatives considered for this section?

I'm not entirely confident that I'd like to invest in this approach, so I'd like to understand why we've eliminated other possible approaches.  For example, why don't we want to put the strategy directly on the lease resource?  Why don't we want to introduce a special object that declares the strategies to use? Why don't we want to make the strategy configurable by an apiserver flag?

Reply:
Added. cc @sttts, hoping we can get consensus on how we want to move forward with this.